File: java/amp-eventing/src/main/java/amp/eventing/EventRegistration.java
package amp.eventing;


import java.util.HashMap;
import java.util.List;
import java.util.Map;

import cmf.bus.Envelope;
import cmf.bus.EnvelopeHeaderConstants;
import cmf.bus.IEnvelopeFilterPredicate;
import cmf.bus.IRegistration;
import cmf.eventing.IEventHandler;


public class EventRegistration implements IRegistration {

    @SuppressWarnings("rawtypes")
	protected IEventHandler eventHandler;
    protected IEnvelopeFilterPredicate filterPredicate;
    protected IInboundProcessorCallback processorCallback;
    protected Map<String, String> registrationInfo;


    @Override
    public IEnvelopeFilterPredicate getFilterPredicate() {
        return filterPredicate;
    }

    @Override
    public Map<String, String> getRegistrationInfo() {
        return registrationInfo;
    }
    
    
    @SuppressWarnings({ "rawtypes" })
    public EventRegistration(IEventHandler eventHandler, IInboundProcessorCallback processorCallback) {

        this.eventHandler = eventHandler;
        this.processorCallback = processorCallback;

        registrationInfo = new HashMap<String, String>();
        registrationInfo.put(EnvelopeHeaderConstants.MESSAGE_TOPIC, eventHandler.getEventType().getCanonicalName());
    }
    
    
	@Override
	@SuppressWarnings("unchecked")
	public Object handle(Envelope env) throws Exception {
		
		Object event = this.processorCallback.ProcessInbound(env);
		Object result = null;
		
		if (null != event) {
            try {
                result = eventHandler.handle(event, env.getHeaders());
            } catch (Exception ex) {
                result = handleFailed(env, ex);
            }
		}

        return result;
    }

    @Override
    public Object handleFailed(Envelope env, Exception ex) throws Exception {
        try {
            return eventHandler.handleFailed(env, ex);
        } catch (Exception failedToFail) {
            throw failedToFail;
        }
    }

    public void setFilterPredicate(IEnvelopeFilterPredicate filterPredicate) {
        this.filterPredicate = filterPredicate;
    }

    public void setRegistrationInfo(Map<String, String> registrationInfo) {
        this.registrationInfo = registrationInfo;
    }
}

File: dotnet/source/amp.eventing/EventRegistration.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Common.Logging;

using cmf.bus;
using cmf.eventing;

using amp.bus;

namespace amp.eventing
{
    public class EventRegistration : IRegistration
    {
        protected Func<IEventHandler<object>, Envelope, object> _interceptor;
        protected IDictionary<string, string> _registrationInfo;
        protected ILog _log;


        public virtual Predicate<Envelope> Filter
        {
            get { return null; }
        }

        public virtual IDictionary<string, string> Info
        {
            get { return _registrationInfo; }
        }

        public virtual IEventHandler<object> Handler { get; protected set; }
        
        public virtual Envelope Envelope { get; protected set; }


        public EventRegistration(
            IEventHandler<object> handler, 
            Func<IEventHandler<object>, Envelope, object> interceptor)
        {
            this.Handler = handler;
            _interceptor = interceptor;

            _registrationInfo = new Dictionary<string, string>();
            _registrationInfo.SetMessageTopic(handler.Topic);

            _log = LogManager.GetLogger(this.GetType());
        }


        public virtual object Handle(Envelope env)
        {
            _log.Debug("Inside RegistrationInterceptor.Handle(Envelope env)");

            return _interceptor(this.Handler, env);
        }

        public virtual object HandleFailed(Envelope env, Exception ex)
        {
            _log.Debug("Enter HandleFailed");

            // either succeed & return the result or fail & return null
            try
            {
                _log.Debug("Leave HandleFailed");
                return this.Handler.HandleFailed(env, ex);
            }
            catch (Exception failedToFail)
            {
                _log.Error("Caught an exception attempting to raise the Failed event", failedToFail);
                return null;
            }
        }
    }
}
